### DISCUSSION TITLE
Integration of SEDS in F´

### DISCUSSION THREAD

**Discussion by ThibFrgsGmz** [nasa/fprime#1173] [Ideas]
| | |
|:---|:---|
|**_F´ Version_**| devel |
|**_Affected Component_**|  F´|
---
## Feature Description

Integrate the SEDS (SOIS-EDS - Spacecraft Onboard Interface Services Electronic Data Sheet) of the CCSDS (Consultative Committee for Spacecraft Data Systems) into the F´ framework.

An EDS [1] is an unambiguous XML representation of the functional interfaces offered by all protocols used to access data interfaces for hardware and software components.

To paraphrase @skliper [2], the main benefits of SEDS are:

- The exchange of unambiguous interface definitions in a standard format between organizations/partners

- Automatic generation of interface code, system models, system tests and mission operational databases

- Runtime support for plug-and-play systems.

## Rationale

According to the reports of the European working group SAVOIR-SAFI [3], the SW-ICDs in proprietary (company-specific) paper format of the unit suppliers will disappear and be homogenized/standardized in electronic format.

An integration of SEDS is still ongoing in the Goddard Space Flight Center cFS framework [4].
The integration of this feature in F´ would be very interesting to promote its use for primes but also for unit suppliers Flight Software.

According to @timcanham, in a discussion on the F' Community Group [5], the modification would not result from an XML change at the level of F´, but from the way we compose the packets that are sent to the Ground System.

## References

> [1] SPACECRAFT ONBOARD INTERFACE SERVICES — XML SPECIFICATION FOR ELECTRONIC DATA SHEETS https://public.ccsds.org/Pubs/876x0b1.pdf

> [2] https://github.com/nasa/cFE/issues/207

> [3] https://hal.archives-ouvertes.fr/hal-01261720/document (Official documents are archived in http://essr.esa.int for registered members.)

> [4] Continuation of EDS integration for CFE: https://github.com/nasa/cFE/issues/243

> [5] F´ Community Group / Integration of SOIS EDS for F´: https://groups.google.com/g/fprime-community/c/I9REPFl3egc

Other interesting links:

> EDS at ESA: Current Status and Roadmap: https://indico.esa.int/event/338/contributions/5707/attachments/4006/5911/1400%20-%20Electronic%20Data%20Sheets.pdf

> SOIS Electronic Data Sheets for Onboard Devices: https://indico.esa.int/event/22/contributions/1997/attachments/1658/1950/SOIS_EDS_ADCSS_2013_v1.0.pdf

**Comment by LeStarch** [comment]
I am going to make this a discussion, as it seems to be a conversation.

**Comment by ThibFrgsGmz** [comment]
Since my post, I have learned that the SEDS standard is part of the International Deep Space Interoperability Standards (https://www.internationaldeepspacestandards.com/) applicable to the software engineering domain (https://nasasitebuilder.nasawestprime.com/idss2/wp-content/uploads/sites/45/2020/10/software baseline final 9-2020.pdf page 13).

I take the liberty of extracting the important elements below, in case the hyperlink is broken in the future.


```
3.2.2.1 CCSDS 876.0-B-1, SPACECRAFT ONBOARD INTERFACE SERVICES--XML
SPECIFICATION FOR ELECTRONIC DATA SHEETS

CCSDS 876.0-B-1, Spacecraft Onboard Interface Services--XML Specification for
Electronic Data Sheets, is a CCSDS Red Book standard specifying the Spacecraft
Onboard Interface Services (SOIS) compliant services to be provided in support of
software applications. This CCSDS standard defines the Extensible Markup Language
(XML) Specification for SOIS Electronic Data Sheet (SEDS) for Onboard Devices.

Definition of the standard encompasses the XML representation of the functional
interfaces offered by protocols used to access the data interfaces. Descriptions in
machine-readable format can be processed by a toolchain to facilitate the various
phases in the life of a space vehicle.

The Dictionary of Terms for Electronic Data Sheets (EDS) specifies the basic format of
the vocabulary, while a publication on SANA contains the actual normative details of the
vocabulary. Additional common terms that apply to all HEOMD projects should be
considered for registration in SANA.

SWS-007
    Spacecraft shall utilize XML specification for electronic data sheets
    standard as defined in CCSDS 876.0-B-1, Spacecraft Onboard Interface
    Services--XML Specification for Electronic Data Sheets.

    Rationale:
        Compliance with the XML-based Electronic Data Sheet (EDS)
        specification provides a means for electronically defining and describing
        data across spacecraft interfaces, with visiting spacecraft, and the ground.
        Definition of the data will allow inter-module processing in support of
        autonomous operations, thereby providing control authority to spacecraft
        software systems.

SWS-008 
    Spacecraft software shall utilize the specification for Dictionary of Terms
    for Electronic Data Sheets (EDS) as defined in CCSDS 876.1-R-2.

    Rationale:
        Compliance with the dictionary of terms for EDS specification
        provides a means for semantic interoperability of data across
        spacecraft. Consistent use of the common semantic attributes and
        quantity types enables the use of data discovery and standardized
        preservation of the relationship between data points. Note: CCSDS 876.1-
        R-2 is currently a CCSDS draft or “Red Book” standard and will become a
        recommended or “Blue Book” standard sometime in the future. Tailoring of
        the standard will be required.
```

These standards are used in NASA's Artemis program, which is supported by several international agencies (ESA, JAXA,...), as well as by other international private partners.

At the ESA MBSE 2021 workshop (https://indico.esa.int/event/386/timetable/), we can see how the large systems integrator Thales Alenia Space has implemented this adherence to the SEDS standard for its Gateway modules (https://indico.esa.int/event/386/contributions/6237/attachments/4276/6467/1715%20-%20Extending%20the%20Scale%20of%20Our%20Modelling%20Environments%20with%20CFS%20and%20SEDS.pdf).

**Comment by ThibFrgsGmz** [comment]
I'm wondering if a framework should include this type of technology stack as well. If not, the mission should be able to incorporate it when using the framework.

If I correctly interpret Uncle Bob's Clean Architecture (https://blog.cleancoder.com/uncle-bob/2012/08/13/the-clean-architecture.html), the SEDS stack would be part of the "Interface Adapters" layer rather than the "Frameworks & Drivers" layer.
However, as long as the domain (the business rules/logic) is self-contained, I don't really see why SEDS couldn't be integrated into a framework. We'd be more akin to hexagonal architecture (https://medium.com/ssense-tech/hexagonal-architecture-there-are-always-two-sides-to-every-story-bc0780ed7d9c), where the infrastructure is located outside of the domain (framework, interface,...).

**Reply by Joshua-Anderson** [reply]
> I'm wondering if a framework should include this type of technology stack as well. If not, the mission should be able to incorporate it when using the framework.

I agree, this seems like an excellent candidate for an external F' library! It seems like this could be implmented as a custom autocoder to generate the necessary XML documentation, and a custom component that replaces the F' packetization and serialization with a SEDS compliant solution.

**Reply by ThibFrgsGmz** [reply]
Since this is an international standard of general interest for interoperability of space software systems, I think we should be more interested in integrating this third-party "SEDS" technology stack rather than developing a custom serialization component.

The implementation of SEDS in a mission must be and remain independent of the framework, just as the generated source code (serialization-encoder/deserialization-decoder) should be framework and domain agnostic since it is shared between the client and the unit/equipment suppliers.

This gives the user the possibility to delay or defer the selection of this type of technology stack when designing the architecture of his software system (adapters, framework, database, OS, HW,...).

This also has the advantage of preserving Uncle Bob's architectural boundary, and the maintenance of the component logic is not the sole responsibility of the framework.

This stack must be open source and used as a reference by all the stakeholders who adhere to this standard, hence the term interoperability.

There are only two "SEDS -> source code" conversion tools that I am aware of:

- EdsLib (https://github.com/nasa/EdsLib) from NASA which is/will be used in cFS, see:
    -  https://github.com/nasa/cFS/discussions/284#discussioncomment-976057
    - https://github.com/nasa/cFS/discussions/307#discussioncomment-1134493)
- SECT - SAVOIR EDS Common Tooling (https://essr.esa.int/project/ccsds-eds-reference-tooling) from ESA, see:
    -  https://github.com/nasa/cFS/discussions/274#discussioncomment-1000438)
